### PR TITLE
Fix #220: EOF after every incremental compilation cycle

### LINKED ISSUE #152: sbt.inc.IncrementalTest disabled
`sbt.inc.IncrementalTest`, which contains a large collection of tests (in an easy to debug harness/format) is currently mostly disabled.

The primary issue is that `sbt.inc.TestAnalysisCallback` (from the same file) has gone out of sync with the private AnalysisCallback in `sbt.inc.internal`. Since the AnalysisCallback impl is completely private, adjusting its implementation to be _directly_ usable for tests would avoid duplication and prevent losing sync again.

### LINKED ISSUE #220: EOF after incremental compilation in sbt
I tried to test newest zinc (based on #216 ) in sbt based on https://github.com/romanowski/sbt/tree/updated-sbt-for-zinc and using my shapless repo stripped of plugins and migrated to 1.0 (https://github.com/romanowski/shapeless/tree/profiling/problem).

Each incremental compilation failed with EOFException (from sbt console):
```
[info] Compiling 1 Scala source to /home/krzysztof/workspace/shapeless/examples/target/scala-2.11/classes... (out-1224)
[info] Done compiling. (out-1224)
Listening for transport dt_socket at address: 5011
[error] (examples/compile:compile) java.io.EOFException: Unexpected end of ZLIB input stream (out-1195)
```

There is some problem with sbt 1.0 logging (`examples/compile:compile`  does not help) however using debugger I was able to extract stack trace:

```
"pool-42-thread-7@17287" prio=5 tid=0x134 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:240)
	  at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158)
	  at java.util.zip.ZipInputStream.read(ZipInputStream.java:194)
	  at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	  at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	  at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	  - locked <0x438c> (a java.io.InputStreamReader)
	  at java.io.InputStreamReader.read(InputStreamReader.java:184)
	  at java.io.BufferedReader.fill(BufferedReader.java:161)
	  at java.io.BufferedReader.readLine(BufferedReader.java:324)
	  at java.io.BufferedReader.readLine(BufferedReader.java:389)
	  at sbt.internal.inc.TextAnalysisFormat$VersionF$.read(TextAnalysisFormat.scala:98)
	  at sbt.internal.inc.TextAnalysisFormat.readCompanionMap(TextAnalysisFormat.scala:85)
	  at sbt.internal.inc.FileBasedStore$FileBasedCompanionsMapStore.$anonfun$getUncaught$2(FileBasedStore.scala:71)
	  at sbt.internal.inc.FileBasedStore$FileBasedCompanionsMapStore$$Lambda$2731.1300274626.apply(Unknown Source:-1)
	  at sbt.io.Using.apply(Using.scala:23)
	  at sbt.internal.inc.FileBasedStore$FileBasedCompanionsMapStore.getUncaught(FileBasedStore.scala:68)
	  at sbt.internal.inc.TextAnalysisFormat$APIsF$.$anonfun$read$9(TextAnalysisFormat.scala:168)
	  at sbt.internal.inc.TextAnalysisFormat$APIsF$$$Lambda$2008.64531322.apply(Unknown Source:-1)
	  at xsbti.SafeLazy$Impl._t$lzycompute(SafeLazy.scala:18)
	  - locked <0x4394> (a xsbti.SafeLazy$Impl)
	  at xsbti.SafeLazy$Impl._t(SafeLazy.scala:16)
	  at xsbti.SafeLazy$Impl.get(SafeLazy.scala:22)
	  at sbt.internal.inc.TextAnalysisFormat$APIsF$.$anonfun$read$11(TextAnalysisFormat.scala:171)
	  at sbt.internal.inc.TextAnalysisFormat$APIsF$$$Lambda$2519.1636850869.apply(Unknown Source:-1)
	  at xsbti.SafeLazy$Impl._t$lzycompute(SafeLazy.scala:18)
	  - locked <0x4396> (a xsbti.SafeLazy$Impl)
	  at xsbti.SafeLazy$Impl._t(SafeLazy.scala:16)
	  at xsbti.SafeLazy$Impl.get(SafeLazy.scala:22)
	  at xsbti.api.AnalyzedClass.api(AnalyzedClass.java:32)
	  at sbt.internal.inc.TextAnalysisFormat$CompanionsF$.$anonfun$write$11(TextAnalysisFormat.scala:190)
	  at sbt.internal.inc.TextAnalysisFormat$CompanionsF$$$Lambda$2730.961927259.apply(Unknown Source:-1)
	  at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
	  at scala.collection.TraversableLike$$Lambda$43.1062186835.apply(Unknown Source:-1)
	  at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:225)
	  at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:432)
	  at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:432)
	  at scala.collection.TraversableLike.map(TraversableLike.scala:234)
	  at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
	  at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	  at sbt.internal.inc.TextAnalysisFormat$CompanionsF$.write(TextAnalysisFormat.scala:190)
	  at sbt.internal.inc.TextAnalysisFormat.writeCompanionMap(TextAnalysisFormat.scala:65)
	  at sbt.internal.inc.FileBasedStore$FileBasedStoreImpl.$anonfun$set$1(FileBasedStore.scala:38)
	  at sbt.internal.inc.FileBasedStore$FileBasedStoreImpl.$anonfun$set$1$adapted(FileBasedStore.scala:31)
	  at sbt.internal.inc.FileBasedStore$FileBasedStoreImpl$$Lambda$2697.344167950.apply(Unknown Source:-1)
	  at sbt.io.Using.apply(Using.scala:23)
	  at sbt.internal.inc.FileBasedStore$FileBasedStoreImpl.set(FileBasedStore.scala:31)
	  at sbt.internal.inc.AnalysisStore$$anon$1.set(AnalysisStore.scala:19)
	  at sbt.internal.inc.AnalysisStore$$anon$2.set(AnalysisStore.scala:30)
	  - locked <0x43a2> (a sbt.internal.inc.AnalysisStore$$anon$2)
	  at sbt.Defaults$.$anonfun$compileTask$1(Defaults.scala:1036)
	  at sbt.Defaults$$$Lambda$1120.2020702385.apply(Unknown Source:-1)
	  at scala.Function1.$anonfun$compose$1(Function1.scala:44)
	  at scala.Function1$$Lambda$915.1531506932.apply(Unknown Source:-1)
	  at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:40)
	  at sbt.internal.util.$tilde$greater$$Lambda$1732.1394524874.apply(Unknown Source:-1)
	  at sbt.std.Transform$$anon$4.work(System.scala:58)
	  at sbt.Execute.$anonfun$submit$2(Execute.scala:232)
	  at sbt.Execute$$Lambda$1746.549801206.apply(Unknown Source:-1)
	  at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	  at sbt.Execute.work(Execute.scala:241)
	  at sbt.Execute.$anonfun$submit$1(Execute.scala:232)
	  at sbt.Execute$$Lambda$1740.191953464.apply(Unknown Source:-1)
	  at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:161)
	  at sbt.ConcurrentRestrictions$$anon$4$$Lambda$1745.4723664.apply(Unknown Source:-1)
	  at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	  at java.lang.Thread.run(Thread.java:745)

```

I think that lazy storing API has serious design problem with file based store since at the time when Zinc wants to store APIs for new analysis, old one does not exists so it can't read APIs from there.

We should either load APIs before deleting old file or write new Analysis to temporary file and later remove old one.

### PR DESCRIPTION
**Pull request by jvican**
This commit introduces the changes proposed in upstream pants to work
around what it happens to be #220, though it wasn't reported back in the
day: https://github.com/pantsbuild/pants/pull/3886/files#diff-297b1b6a94d202511b784fd43284b6ebR119.

As it reads in the original PR, these changes ensure that when analysis
update fails, the analysis file remains intact.

I think it also makes this operation safe to data races, though I have
seen some use of `synchronized` in adjacent pieces of code, so I'm not
really sure if this is true. Comments on this welcome.

The reason why we don't use `java.nio.file.Files` is because of
compatibility issues. Having all the Zinc infrastructure relying on IO
utils, I didn't want to force users of zinc to use JDK versions > 6 (NIO
is not present otherwise). The current implementation of `IO.move` is a
little bit more verbose and it has some cycles (see implementation of
`createDirectory`, for instance), but I think it's reasonable enough to
depend on it and don't reinvent the wheel.

I have not yet double-checked whether this issue fixes #220, so I'll
follow up on this PR soon to confirm this is the case and proceed with
the review.

### CODE REVIEW

**Comment by jvican** [comment]
Agrr, `hub` cached for some reason the previous PR description and it changed the title. I have renamed it to describe the actual fix.

**Comment by jvican** [comment]
Review by @romanowski.

**Comment by romanowski** [comment]
There is small problem with testing that since sbt is incompatible with current zinc (https://github.com/sbt/sbt/pull/2955). I will try to rebase on some old commit zinc.

**Comment by romanowski** [comment]
Tested and seems to work!

@jvican @peiyuwang Thanks!

**Comment by jvican** [comment]
Great, I guess this is ready for review. I've launched again CI. /cc @eed3si9n

File: internal/zinc-persist/src/main/scala/sbt/internal/inc/FileBasedStore.scala [hunk 1]
```diff
@@ -42,8 +49,10 @@ object FileBasedStore {
             outputStream.closeEntry()
           }
       }
+      IO.move(tmpAnalysisFile, file)
```
Comment (peiyuwang) [inline internal/zinc-persist/src/main/scala/sbt/internal/inc/FileBasedStore.scala:52]: there is a small risk file can still be corrupted, note `IO.move` does file copy if `renameTo` fails, and file copy may fail

https://github.com/sbt/io/blob/4377f594e1d26d0486792e9ac6dce2c9d2cd9133/io/src/main/scala/sbt/io/IO.scala#L796

**Review by eed3si9n — APPROVED** [review]

**Comment by eed3si9n** [comment]
Thanks guys for fixing this and wrapping it up in a PR.

File: internal/zinc-persist/src/main/scala/sbt/internal/inc/FileBasedStore.scala (hunk 1 above)
Comment (jvican) [reply internal/zinc-persist/src/main/scala/sbt/internal/inc/FileBasedStore.scala:52]: Good point, I'll follow up with a new PR, though I'm not sure if there's an equivalent for `REPLACE_EXISTING` in JKD 6.

**Pull request merged**